### PR TITLE
Fix pipe close operation so it doesn't break the other side of pipe

### PR DESCRIPTION
**Pull request by hoodmane**
Without this PR, the added test fails when we try to read from the read end of the pipe:
```
if(read(fdr, buf, 5) != 4){
```
because 
https://github.com/emscripten-core/emscripten/blob/88fc9f0ea67aa7be76b2d482476ad3c08ab3cbb3/src/library_pipefs.js#L218
sets the pipe buckets to null and then `read` tries to access `pipe.buckets.length` here:
https://github.com/emscripten-core/emscripten/blob/88fc9f0ea67aa7be76b2d482476ad3c08ab3cbb3/src/library_pipefs.js#L90
By contrast, the test works fine on an ordinary posix machine. I just modified the logic to only delete `pipe.buckets` when both ends of the pipe are closed.

### CODE REVIEW

**Comment by hoodmane** [comment]
Well this is not quite the right approach. Instead, we should just make `end1.pipe` and `end2.pipe` different objects. Otherwise it doesn't work if we `dup2` one of the ends and then close it.

**Comment by hoodmane** [comment]
Okay I switched to storing a refcnt, the `dup` operations can just increment the refcount (once they are fixed to work on pipes at all).

**Comment by hoodmane** [comment]
The `test-other` job didn't run my `test_fd_pipe.c` test. Is there a manifest somewhere I have to add it to?

**Comment by hoodmane** [comment]
Is there something I have to do to get the test to compile? I don't see anything else for `test_fd_closed.cpp`.

File: src/library_pipefs.js [hunk 1]
```diff
@@ -215,7 +216,10 @@ mergeInto(LibraryManager.library, {
       },
       close: function (stream) {
         var pipe = stream.node.pipe;
-        pipe.buckets = null;
+        pipe.refcnt --;
+        if(pipe.refcnt === 0){
```
Comment (kripken) [inline src/library_pipefs.js:220]:
```suggestion
        if (pipe.refcnt === 0){
```

File: tests/other/test_fd_pipe.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 The Emscripten Authors.  All rights reserved.
```
Comment (kripken) [inline tests/other/test_fd_pipe.c:1]:
```suggestion
// Copyright 2021 The Emscripten Authors.  All rights reserved.
```

**Comment by kripken** [comment]
> The test-other job didn't run my test_fd_pipe.c test. Is there a manifest somewhere I have to add it to?

Jobs need to be described in `test/test_*.py`. In this case, `tests/test_other.py`. But it looks like you've done that. And I do see the test in the latest CI output.

**Review by kripken — COMMENTED** [review]
I'd need to read this more carefully to understand why 2 is the right number for the ref count - a comment would be good there. Or is it obvious?

cc @robertaboukhalil who might know this code.

**Comment by robertaboukhalil** [comment]
> I'd need to read this more carefully to understand why 2 is the right number for the ref count - a comment would be good there. Or is it obvious?
> 
> cc @robertaboukhalil who might know this code.

Hmm, I'm not familiar with PIPEFS, but is it 2 because there's 1 for read, 1 for write, and you want it to error out when both have been closed?

**Comment by hoodmane** [comment]
Thanks for the review @kripken!

> why 2 is the right number for the ref count - a comment would be good there

A comment would be good there. It's because there are two ends to the pipe.

**Comment by hoodmane** [comment]
The important thing is that it must be possible to read from the read end of the pipe even after the write end is closed.

File: tests/other/test_fd_pipe.c [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+int pipefd[2];
+char buf[5];
+
+int main() {
+  const char *filename = "test.dat";
+
+  // Create a pipe
+  int status = pipe(pipefd);
+  if (status == -1) {
```
Comment (sbc100) [inline tests/other/test_fd_pipe.c:22]: You can just use assert() here and elsewhere in this file rather then returning codes.   That way when it fails the test will show exactly which line it was that failed.

File: tests/test_other.py [hunk 4]
```diff
@@ -7810,6 +7810,9 @@ def test_ioctl_window_size(self):
   def test_fd_closed(self):
     self.do_other_test('test_fd_closed.cpp')
 
+  def test_fd_pipe(self):
+    self.do_other_test('test_fd_pipe.c')
```
Comment (sbc100) [inline tests/test_other.py:7814]: We already have several pipe tests.. could just just modify `tests/unistd/pipe.c` instead perhaps?

**Comment by hoodmane** [comment]
Thanks @sbc100, I tried to address your comments.

File: src/library_syscall.js [hunk 5]
```diff
@@ -374,7 +378,10 @@ var SyscallsLibrary = {
   },
   __sys_dup: function(fd) {
     var old = SYSCALLS.getStreamFromFD(fd);
-    return FS.open(old.path, old.flags, 0).fd;
+    if(old.node.pipe){
```
Comment (sbc100) [inline src/library_syscall.js:381]: add space after `if`

File: src/library_syscall.js [hunk 6]
```diff
@@ -137,10 +137,14 @@ var SyscallsLibrary = {
       }
       return 0;
     },
-    doDup: function(path, flags, suggestFD) {
+    doDup: function(stream, suggestFD) {
       var suggest = FS.getStream(suggestFD);
       if (suggest) FS.close(suggest);
-      return FS.open(path, flags, 0, suggestFD, suggestFD).fd;
+      if(stream.node.pipe){
```
Comment (sbc100) [inline src/library_syscall.js:143]: space after `if` and before `{`

File: system/lib/dlmalloc.c [hunk 7]
```diff
@@ -32,7 +32,7 @@
   Emscripten aligns even float128 to 64-bits, to save size and increase speed.
   See https://github.com/emscripten-core/emscripten/issues/10072
 */
-_Static_assert(MALLOC_ALIGNMENT == 8, "max_align_t must be 8");
+// _Static_assert(MALLOC_ALIGNMENT == 8, "max_align_t must be 8");
```
Comment (sbc100) [inline system/lib/dlmalloc.c:35]: Looks like you need to get the latest version of llvm (try `emsdk install tot`)

Comment (hoodmane) [reply system/lib/dlmalloc.c:35]: Oops didn't mean to commit this =)

File: src/library_syscall.js (hunk 6 above)
Comment (hoodmane) [reply src/library_syscall.js:143]: To be clear, I assume that there isn't an automatic formatter that won't choke on the `#ifdef`'s and stuff?

Comment (sbc100) [reply src/library_syscall.js]: I think `clang-format` can operate on just a single line or range of lines in your editor.. although I don't use it myself for JS much.

**Comment by kripken** [comment]
Code looks good to me. Most of the test errors here are old errors on main, so merging latest to here should clear them up. However, it looks like the pipe test itself fails, so that needs to be fixed I think.

**Comment by kripken** [comment]
`test_unistd_pipe` - runnable locally with `tests/runner.py test_unistd_pipe`.

**Comment by hoodmane** [comment]
> the pipe test itself fails

I guess I didn't test the updated version after addressing @sbc100's comments. I find it most convenient to test the test with `gcc`. Testing the gcc behavior is probably a good idea anyways for all tests that aren't intended to involve emscripten-specific functionality or behaviors.

Should be fixed now.

**Comment by hoodmane** [comment]
Okay, now `gcc` accepts it but `clang` is still upset.

Also, what is the suggested way to configure `emsdk` to use a local development copy of `emscripten`? Is there a spot in the contributing docs that explains this?

**Comment by hoodmane** [comment]
`tests/runner.py test_unistd_pipe` now works locally for me.

File: tests/unistd/pipe.c [hunk 8]
```diff
@@ -136,6 +136,30 @@ int main()
...
+    assert(close(fd[1]) == 0);
+    // This write should fail
+    assert(write(fd[1], "YYYY", 4) == -1);
+    // The error number is EBADF
+    assert(errno == EBADF);
+
+    // read from the other end of the pipe
+    assert(read(fd[0], buf, 5) == 4);
+    // We should have read what we wrote to the other end
+    assert( strncmp((const char *)buf, "XXXX", 5) == 0);
```
Comment (sbc100) [inline tests/unistd/pipe.c:154]: Remove space after open paren.

Comment (sbc100) [reply tests/unistd/pipe.c:154]: Maybe use `memcmp` to avoid the cast?

**Review by sbc100 — COMMENTED** [review]
Great!  lgtm.

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
Thanks @hoodmane !

**Pull request merged**